### ISSUE TITLE
[Feature request] Integration with image translators

### ISSUE THREAD

**Issue by EnergoStalin** [ppixiv/ppixiv#79]
I'd like to have integration with [cotrans.touhou.ai](https://cotrans.touhou.ai) or any self hosted alternative like [this](https://github.com/zyddnys/manga-image-translator) for example(which cotrans use under the hood btw).

Cotrans already have [userscript](https://greasyfork.org/en/scripts/437569-cotrans-manga-image-translator-regular-edition) for pixiv and several other sites. But extending their script with ppixiv support seems not the most optimal solution due to your UX and lack of documented ppixiv api's for doing it(they also dont have documented api tho).

So i'd like to hear you thoughts about how you think it can be implemented and will you accept it as core functionality(making image translation provider pluggable of course because we don't want depend on third parties do we?) or expose some api's to extend ppixiv that way.

This is the last thing that forces me switching on pixiv from time to time and i'm finally sick of it. So maybe i considering a PR later or write a separate plugin if there's some work for community left).

**Comment by EnergoStalin** [comment]
What i see from my perspective.

Poorly designed(~20 min)

## API part
### Core solution
If make it core functionality expose several endpoints to register providers which may look like this
```ts
type id = string;

interface TranslationOptions {
    language: string
    resolution: string
    detector: string
    direction: string
    translator: string
}

interface ImageTranslationProvider {
    status(id: id): string
    download(id: id): Promise<Blob>
    upload(data: Blob, options: TranslationOptions): id
}
```

### API solution
In case of exposing only api's somehow allow extending settings page and add a hook on loading image set(you better know what it means).

## Config
### Core solution
Make extra top level tab for this functionality with settings like cotrans do they're the same(not sure) as aforementioned alternative
![image](https://github.com/ppixiv/ppixiv/assets/49844515/25f3296f-8d42-4be0-99c2-22802c0387dd)

Also maybe introduce keybinds for that.

### API solution
Expose api to add settings tab you also know better

## UX part
### Core solution
Guess it's better do like cotrans extension does on illustration page like extra button on info popup displaying progress in same popup or in the empty right corner
### API solution
Provide api's for extending popup menu and accessing to raw images

**Comment by ppixiv** [comment]
I wonder if that service will stick around for very long.  It looks like somebody's side project and it costs money to run (sometimes things like that stick around forever, sometimes they don't...).  But we can give it a try, there's already support for image overlays in the viewer so it's not hard to add.

I've made a pass at it.  It's not in a release yet, but you can try it with the beta build:

https://ppixiv.org/test

It doesn't have settings yet and is just enabled from the context dropdown menu.  It won't work on iOS (and probably not MacOS) because of Safari restrictions.  It's pretty slow since we have to redownload images even if they were just downloaded for viewing, since Pixiv doesn't have CORS set up for its own images.  It uses Pixiv's lower-resolution images to make that re-downloading a lot faster, but that means the text is lower res too.  That can be an option when settings are added.

**Comment by EnergoStalin** [comment]
> I wonder if that service will stick around for very long.

That's why i suggested self hosted alternatives support with plugable translator it would be tricky but definitely possible either through `// @connect *` or `// @connect localhost` with proxy which is the best option can be offered in that case. Cotrans can be default option but making it the only need to be thoughtfully considered. Most people like myself would be totally okay until it works tho.

> I've made a pass at it. It's not in a release yet

Works perfectly! Except mobile there's simply no option for that. Yet? Waiting for implemented options & mobile support. Also would be great having translation status reflected on
![image](https://github.com/ppixiv/ppixiv/assets/49844515/7b7c1fc1-bd49-4443-9f02-ac1a67e6efef)
too because i noticed sometimes translation can stuck. Or any other option like timeout if this would be easier to implement cause translation itself rarely run longer than 15-20 seconds after upload.

> It won't work on iOS (and probably not MacOS) because of Safari restrictions.

That's what i did not expect to hear from someone who stick build system to windows only). But that's just means they should use chrome or alternatives. Just like i need switch to windows for building it as is).

p.s. No complains great work so far

> It's pretty slow since we have to redownload images

Don't think slow speed is a problem in case of good connection. The higher resolution the better for text recognition and for peoples eyes. Especially having low resolution in case of monitor > 1080p would be a pain.

Didn't notice they uses inpaint masks as overlay at first which makes it lesser a problem or not at all. Let's use you solution for now.

My brain whole time think'ed it render masks on top of images and send it back that way. Because i wrote [script](https://github.com/EnergoStalin/cotrans) utilizing manga-image-translator and google collab for mass translation files which do so.

> That can be an option when settings are added.

Yeah it might depend on Detection Resolution option download full images or not(of course if final resolution known before downloading). As i know cotrans userscript download full images in any case btw.

## Some other edge cases worth mentioning
1. Can be blank translations using gpt translator most reliable is google which i always use [example](https://www.pixiv.net/ranking.php#ppixiv?view=illust&illust_id=111797682&page=2).
 
    Mostly gpt refuses translate what don't like in this case i guess it's topic fearing of man's. Also on early stages of using that service i encountered what they just run out of paid tokens.
2. Translation stuck because of api error

    Rarely experience this but still can be a problem. Mostly relatable to custom providers which i suggested.
3.  Because of previous 2, requeueing single pages with different settings is must have option also not all images have text either so image picker would be handful but it's out of scope

**Comment by EnergoStalin** [comment]
There's a official cotrans [client](https://www.npmjs.com/package/@cotrans/client?activeTab=readme) btw didn't know it exists. Discovered it in this chineese [topic](https://greasyfork.org/en/scripts/437569-cotrans-manga-image-translator-regular-edition/discussions/192031). Currently digging into this because still no examples thanks for jsdoc tho

**Comment by EnergoStalin** [comment]
Also how about saving images with translation rendered on top of it? This wont be easy tho cause i guess images need to be redownloaded again. As possible solution i can suggest using URL.createObjectURL in illust-view widget on original images and/or thumbnails with something like [that](https://stackoverflow.com/questions/11876175/how-to-get-a-file-or-blob-from-an-object-url) to get it back, it has wide support in browsers and you already using it in some cases so i don't think you would be against that too much. Or does this method has other limitations i don't know about? The only problem i can see what it increase complexity and potentially could harm stability due to current navigation schema which require manual memory management for such urls.

Current solution performs caching well but if we need to do something with illustrations it becomes problematic and lead to unnecessary traffic as i can see.

**Comment by ppixiv** [comment]
I don't think many people would actually run their own server if the translation service goes away (it's not packaged to be easy for regular users to set up, it needs setting up API keys, etc).  I've added a setting to point it at another URL though, you can change it from the console:

ppixiv.settings.set("translation_api_url", "url");

> That's what i did not expect to hear from someone who stick build system to windows only). But that's just means they should use chrome or alternatives. Just like i need switch to windows for building it as is).

iOS doesn't have alternative browsers, all browsers on iOS are just wrappers around Safari's backend.  Apple doesn't allow other browser engines.  I think this almost works now, but "Stay" has a bug in GM.xmlHttpRequest that prevents it from working anyway (it doesn't handle FormData and just sends "[object Object]").  Also, "Stay" doesn't actually check @connect, it just lets you connect anywhere, which isn't greaet.  Script managers on iOS aren't in a great place...

The build system isn't Windows-specific, it's just never been tested anywhere else.  I poked at it and it only needed some SASS fixes.  VVpython and the other native stuff won't work, but that isn't needed just to build.

> Don't think slow speed is a problem in case of good connection. The higher resolution the better for text recognition and for peoples eyes. Especially having low resolution in case of monitor > 1080p would be a pain.

Pixiv is just slow in the US, the re-downloading can be 50% or more of the total wait time.  I've made it use the CF mirror if it's enabled which helps speed it up, but it sometimes causes connection problems so it's not on by default.

> ## Some other edge cases worth mentioning
> 1. Can be blank translations using gpt translator most reliable is google which i always use [example](https://www.pixiv.net/ranking.php#ppixiv?view=illust&illust_id=111797682&page=2).

It works fine with gpt3.5 with the low res image, it's only blank with the full image.  Probably a bug in the translation server?

>    Mostly gpt refuses translate what don't like in this case i guess it's topic fearing of man's. Also on early stages of using that service i encountered what they just run out of paid tokens.
> 2. Translation stuck because of api error
>    Rarely experience this but still can be a problem. Mostly relatable to custom providers which i suggested.
> 3. Because of previous 2, requeueing single images with different settings is must have option also not all images have text either so image picker would be handful but it's out of scope and mostly not worth efforts

I'll think on it, the UI flow isn't obvious and I want to keep the UI simple.

> Also how about saving images with translation rendered on top of it?

It should work in the current build.  It's not great (no download queue, etc.), building out a real download UI doesn't seem worth it with the CORS problem causing double-downloading every time.

**Comment by EnergoStalin** [comment]
> it's not packaged to be easy for regular users to set up, it needs setting up API keys, etc

Not quite true API keys not required for google tranlator. GPU passthrough to docker may be tricky especially on windows but for cpu docker setup is trivial. Not mentioning google colab hosting possibility(it's against their TOS btw but still achievable) [here's](https://colab.research.google.com/drive/1qcllhfCuT5zVRd-7lPvJvE7LHQIi2NJn?usp=sharing) the setup. It will require api client like [this](https://github.com/EnergoStalin/cotrans/blob/41e78e349fd99469593a3fa64ce375b1734f0207/index.js#L25-L62) and http client like [this](https://github.com/EnergoStalin/cotrans/blob/41e78e349fd99469593a3fa64ce375b1734f0207/index.js#L9-L12) for accessing it through localit tunnel

> ppixiv.settings.set("translation_api_url", "url");

Good need to be documented then. Also cotrans and manga-image-translator api not compatible as is. Requests seems similar tho unlike routes.

> Script managers on iOS aren't in a great place...

True. Luckily i have zero mac devices) 

> The build system isn't Windows-specific, it's just never been tested anywhere else.

Yes that's why i said `as is`. Maybe some build system changes let github actions take place more easily.

> I've made it use the CF mirror if it's enabled which helps speed it up, but it sometimes causes connection problems so it's not on by default.

👍🏻 Yeah mirror helps sometimes especially when pixiv struggles.

> It works fine with gpt3.5

It MOSTLY works fine but the point is what it may require retrying with different translator because of mentioned earlier
![](https://user-images.githubusercontent.com/49844515/269161739-b54a1406-798c-49cf-afa5-5eab2dfd13c9.png)
possible GPT schizo behaviour. GPT isn't most reliable technology cause it trying to simulate human behaviour)

>  Probably a bug in the translation server?

Maybe so but i hope mostly you can make it work tricking settings a little.

> I'll think on it, the UI flow isn't obvious and I want to keep the UI simple.

Great goal. I used ComicReader a little they just have toasts(i guess?) in case of errors and global toggle to ignore translation cache. This might be simple enough.

> It should work in the current build

👍🏻 Single image download just fine batch translation is work for another tool. I have [script](https://github.com/EnergoStalin/cotrans/blob/master/index.js) for local files massive translation on colab for this.

# Remained questions
- Still can't see toggle for mobile is it planned or i'm just blind?
- [Part with possible redownloading solution ignored?](https://github.com/ppixiv/ppixiv/issues/79#issuecomment-1730872601)

**Comment by EnergoStalin** [comment]
Can confirm translation working if
![image](https://github.com/ppixiv/ppixiv/assets/49844515/fbbecb2a-9e4d-40ff-bdba-9890ae05b21f)
allow it.

Tested on com.yandex.browser.alpha (116.0.5845.41)

**Comment by ppixiv** [comment]
I tried testing yandex, but it didn't work on my Android tablet (loads but none of the UI responds), so I can't test.  I don't know if there are other browsers that support user scripts on Android, last time I looked I didn't find any support for them on Android.  I've switched it to just disable on iOS for now.

I filed a bug asking to add CORS to the API which would fix iOS, but the cotrans repo is pretty idle so I don't know if I'll get a reply.  It would fix the iOS problems, and also avoid problems with Chrome's nasty throttling.  The API returns 500 errors a lot, and if that happens a few times in a row Chrome blocks access to it for several minutes.  CORS would fix this since the API call would come from the website instead of the browser extension.

I made a first pass at settings.  The translation settings tab has default settings.  Clicking the gear icon in the context menu for an image opens settings for the current image.  These only affect the current page and not all pages for manga, so fiddling with translation settings for a problematic page doesn't cause it to retranslate all pages every time something is changed.  It's not perfect (the settings page covers up the image) but I don't want to do a complex custom UI for it right now.  Failed translations can be retried from here, though I haven't run into any recoverable errors (any time it fails it doesn't work if I try again).

Low res images is disabled for now.  There's a bug in the translation API, if you send the same image in low-res and high-res it thinks they're the same file and returns the cached image, even though the resolution has changed.  Not sure if that's a problem in cotrans or manga-image-translator.

> Part with possible redownloading solution ignored

(I can't reply sentence by sentence to two page replies...)  You can't use createObjectURL on an image, it's for getting a URL for a blob.  There's no way to access the image data without using GM.xmlHttpRequest since the site itself doesn't have access to them, GM.xmlHttpRequest bypasses that but seems to not share cache with the site.

I've temporarily put translations behind a flag, since I want to give the cotrans guy a chance to reply.  If CORS is set up then I might not need to add it \@connect.  To turn it on, run:

ppixiv.settings.set("dev", true);

and add cotrans.touhou.ai and i-cf.pximg.net to the connect list in the script manager settings if needed.  (It's strange that many script managers don't even implement \@connect and just let scripts access anything, maybe only TamperMonkey actually does...)

**Comment by EnergoStalin** [comment]
>  The translation settings tab has default settings. Clicking the gear icon in the context menu for an image opens settings for the current image.

Looks good i don't even think it would need more ui than implemented.

> Low res images is disabled for now. There's a bug in the translation API, if you send the same image in low-res and high-res it thinks they're the same file and returns the cached image, even though the resolution has changed. Not sure if that's a problem in cotrans or manga-image-translator.

Okay waiting it's not critical tho cause you can just change resolution definition to force rescanning. But only 3 times.

> You can't use createObjectURL on an image

The flow i suggested is regularly download image and place it's data to ObjectURL then you can get it whenever you want by that url(cache should be cached here) and set it as img src but it will require downloading all images through such ObjectURL proxy for consistency if i correctly understand how it works.

[Look](https://github.com/hymbz/ComicReadScript/blob/bda82dabae5314dde66857bfdf7009b8752c8997/src/site/jm.tsx#L96) this chineese guy somehow do it not sure is this exactly what we need tho.
![image](https://github.com/ppixiv/ppixiv/assets/49844515/8ad7fb87-35a2-45f2-ad65-4f6187bfcc46)

But if you said redownload would not longer be needed if cotrans introduce cors then be it.

> I've temporarily put translations behind a flag, since I want to give the cotrans guy a chance to reply

Okay luckily i finally learned how to use mobile browser debugging and don't need separate script to set it up.

Translations still working for me here's the non informative video https://github.com/ppixiv/ppixiv/assets/49844515/51a9f83f-9915-4337-a1ab-03d216620658
Except this black screen cant reproduce it anymore tho.

Manga: https://www.pixiv.net/ranking.php?mode=daily#ppixiv?illust_id=112126014&view=illust

**Comment by EnergoStalin** [comment]
So i belive there's only CORS(redownload) problem left for now. Waiting for the release

**Comment by ppixiv** [comment]
I've enabled it in r229.  It's not in the \@connect list, instead I set "\@connect *" so the user's script manager will ask for permission the first time it's used.  That way it'll still work if it's pointed to a different server URL without having to manually add to the connect list.  I think only TamperMonkey bothers with \@connect anyway.

If manga-image-translator becomes easier to set up I could look at adding support for that API, but currently it seems like too much of a hassle to set up so I doubt anyone would use it.  If there was a simple standalone installer it would be more approachable, but a Docker-based server app isn't meant for end users.  I could have VView load it (the local server backend for local image browsing, currently only for Windows), but I don't think I'd use it enough to want to maintain that sort of integration.  manga-image-viewer being GPL is a turn-off too, but not a dealbreaker if it's kept isolated.  Maybe I'll revisit it if something happens to cotrans.
